File: frontend/coordinator-android/app/src/main/java/ug/qaat/coordinator/ui/TimetableGrid.kt
package ug.qaat.coordinator.ui

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.time.LocalDate

/**
 * The weekly timetable, drawn once and shared by every role that shows one.
 *
 * The coordinator's dashboard and the student's home screen used to draw the week two different
 * ways — a Time × Day grid on one, a list of weekday headings on the other — so the same week
 * looked like two different timetables depending on who was holding the phone. Both now render
 * through [TimetableGrid]; only the mapping into [TtEntry] differs.
 */
internal val DAYS = arrayOf("", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")

/** One block on the grid, whatever screen it came from. */
internal data class TtEntry(
    val name: String,
    val dayOfWeek: Int,          // 1=Mon … 7=Sun
    val start: String,           // "HH:MM"
    val durationMin: Int,
    /** Room · lecturer · phone — whatever the caller has; blank parts are dropped. */
    val detail: String = "",
)

/**
 * Time × Day grid: a Time column on the left, day columns across, and each entry rendered as a
 * block COVERING all the cells for its running time (08:00–11:00 = a 3-hour tall block).
 *
 * [sessionType] picks the columns — a weekend cohort gets Sat/Sun instead of Mon–Fri. Entries
 * falling outside the chosen columns are still shown, under "Not on this week's days", so a slot
 * can never silently vanish because its cohort was mis-tagged.
 *
 * [days] overrides that choice outright, for the one caller whose week is not a cohort's week: a
 * LECTURER teaches across cohorts, so their grid must run Monday to Sunday. Deriving that from a
 * session type would mean inventing a fake one, and the next reader would have to work out what
 * "weekend" meant on a screen showing five weekdays.
 */
@Composable
internal fun TimetableGrid(
    entries: List<TtEntry>,
    sessionType: String,
    unscheduledNames: List<String> = emptyList(),
    days: List<Int>? = null,
) {
    @Suppress("NAME_SHADOWING")
    val days = days ?: if (sessionType.lowercase().contains("weekend")) listOf(6, 7) else listOf(1, 2, 3, 4, 5)
    val scheduled = entries.filter { it.dayOfWeek in days && it.start.isNotBlank() }
    val offDays = entries.filter { it.dayOfWeek !in days && it.start.isNotBlank() }
    val today = LocalDate.now().dayOfWeek.value

    var lo = 8; var hi = 19
    scheduled.forEach {
        val h = ttHourOf(it.start)
        if (h < lo) lo = h
        if (h + ttSpan(it.durationMin) > hi) hi = h + ttSpan(it.durationMin)
    }
    lo = lo.coerceIn(0, 23); hi = hi.coerceIn(lo + 1, 24)   // clamp to a real clock — no 24:00/25:00 rows
    val hours = (lo until hi).toList()
    val rowH = 40.dp
    val dayW = 108.dp
    val timeW = 46.dp
    val brand = MaterialTheme.colorScheme.primary
    val line = MaterialTheme.colorScheme.outlineVariant
    val muted = MaterialTheme.colorScheme.onSurfaceVariant

    Column {
        Row(Modifier.horizontalScroll(rememberScrollState())) {
            // Time column
            Column {
                Box(Modifier.width(timeW).height(26.dp), contentAlignment = Alignment.Center) {
                    Text("Time", fontSize = 10.sp, fontWeight = FontWeight.Bold, color = brand)
                }
                hours.forEach { h ->
                    Box(Modifier.width(timeW).height(rowH), contentAlignment = Alignment.TopCenter) {
                        Text(ampmHour(h), fontSize = 9.sp, color = muted, modifier = Modifier.padding(top = 2.dp))
                    }
                }
            }
            // Day columns
            days.forEach { d ->
                val isToday = d == today
                Column {
                    Box(
                        Modifier.width(dayW).height(26.dp)
                            .background(if (isToday) brand else MaterialTheme.colorScheme.surfaceVariant),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text(
                            DAYS[d] + if (isToday) " •" else "", fontSize = 10.sp, fontWeight = FontWeight.Bold,
                            color = if (isToday) MaterialTheme.colorScheme.onPrimary else muted,
                        )
                    }
                    Box(Modifier.width(dayW).height(rowH * hours.size.toFloat())) {
                        // hour gridlines
                        Column {
                            hours.forEach { _ ->
                                Box(Modifier.fillMaxWidth().height(rowH).drawBehind {
                                    drawLine(line, Offset(0f, size.height), Offset(size.width, size.height), 1f)
                                })
                            }
                        }
                        // blocks — height ∝ duration so they cover their full time
                        scheduled.filter { it.dayOfWeek == d }.forEach { e ->
                            val sh = ttHourOf(e.start); val span = ttSpan(e.durationMin)
                            Box(
                                Modifier.offset(y = rowH * (sh - lo).toFloat())
                                    .height(rowH * span.toFloat()).fillMaxWidth().padding(1.5.dp),
                            ) {
                                Surface(
                                    color = MaterialTheme.colorScheme.surface, shape = RoundedCornerShape(5.dp),
                                    border = BorderStroke(1.dp, brand), modifier = Modifier.fillMaxSize(),
                                ) {
                                    Column(Modifier.padding(horizontal = 4.dp, vertical = 3.dp)) {
                                        Text(e.name, fontWeight = FontWeight.SemiBold, fontSize = 10.sp,
                                            lineHeight = 11.sp, maxLines = 2, color = brand)
                                        Text(timeRange(e.start, e.durationMin), fontSize = 8.sp,
                                            lineHeight = 10.sp, color = muted, maxLines = 1)
                                        if (e.detail.isNotBlank()) {
                                            Text(e.detail, fontSize = 8.sp, lineHeight = 10.sp, color = muted,
                                                maxLines = if (span > 1) 3 else 1)
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        if (offDays.isNotEmpty()) {
            Text("Not on this week's days", fontWeight = FontWeight.Bold, fontSize = 12.sp,
                color = MaterialTheme.colorScheme.tertiary, modifier = Modifier.padding(top = 12.dp, bottom = 4.dp))
            offDays.forEach {
                Text("• ${it.name} — ${DAYS.getOrNull(it.dayOfWeek).orEmpty()} ${it.start}".trim(),
                    fontSize = 12.sp, color = muted)
            }
        }
        if (unscheduledNames.isNotEmpty()) {
            Text("Not yet scheduled", fontWeight = FontWeight.Bold, fontSize = 12.sp,
                color = MaterialTheme.colorScheme.tertiary, modifier = Modifier.padding(top = 12.dp, bottom = 4.dp))
            unscheduledNames.forEach { Text("• $it", fontSize = 12.sp, color = muted) }
        }
    }
}

internal fun ttHourOf(hhmm: String) = hhmm.split(":").getOrNull(0)?.toIntOrNull() ?: 0

internal fun ttSpan(mins: Int): Int { val m = if (mins <= 0) 60 else mins; return maxOf(1, (m + 59) / 60) }

internal fun ampmHour(h: Int): String {
    val hh = ((h % 24) + 24) % 24
    val suffix = if (hh < 12) "AM" else "PM"
    val h12 = if (hh % 12 == 0) 12 else hh % 12
    return "$h12 $suffix"
}

/** "14:30" → "2:30 PM". */
internal fun ampm(hhmm: String): String {
    if (hhmm.isBlank()) return ""
    val p = hhmm.split(":"); if (p.size < 2) return hhmm
    val h = p[0].toIntOrNull() ?: return hhmm
    val suffix = if (h < 12) "AM" else "PM"
    val h12 = when { h % 12 == 0 -> 12; else -> h % 12 }
    return "$h12:${p[1]} $suffix"
}

/** "09:00" + 120 → "9:00 AM – 11:00 AM". */
internal fun timeRange(start: String, mins: Int): String {
    if (start.isBlank()) return ""
    if (mins <= 0) return ampm(start)
    val p = start.split(":"); if (p.size < 2) return ampm(start)
    val total = (p[0].toIntOrNull() ?: return ampm(start)) * 60 + (p[1].toIntOrNull() ?: return ampm(start)) + mins
    val end = "%02d:%02d".format((total / 60) % 24, total % 60)
    return "${ampm(start)} – ${ampm(end)}"
}

File: frontend/coordinator-flutter/lib/ui/coordinator/timetable_grid.dart
import 'dart:math';

import 'package:flutter/material.dart';

const List<String> kDays = ['', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

/// One booked block of the grid.
class TtEntry {
  const TtEntry({
    required this.name,
    required this.dayOfWeek,
    required this.start,
    required this.durationMin,
    this.detail = '',
  });

  final String name;
  final int dayOfWeek;
  final String start;
  final int durationMin;
  final String detail;
}

/// A time × day grid of a week's lectures — the shared timetable the coordinator's
/// Home tab and the lecturer's Timetable both render. Coordinate conventions match
/// the native `TimetableGrid`: day 1=Mon…7=Sun, start is "HH:MM", and a unit runs
/// from its start for `durationMin` minutes.
class TimetableGrid extends StatelessWidget {
  const TimetableGrid({
    super.key,
    required this.entries,
    this.sessionType = '',
    this.unscheduledNames = const [],
    this.days,
  });

  final List<TtEntry> entries;
  final String sessionType;
  final List<String> unscheduledNames;

  /// Which day columns to show. When omitted: weekends for a "Weekend" session
  /// type, Monday→Friday otherwise. The lecturer passes all seven explicitly.
  final List<int>? days;

  static const double _rowH = 40;
  static const double _dayW = 108;
  static const double _timeW = 46;

  int _hourOf(String hhmm) {
    final i = int.tryParse(hhmm.split(':').first);
    return i ?? 0;
  }

  int _span(int mins) {
    final m = mins <= 0 ? 60 : mins;
    return max(1, (m + 59) ~/ 60);
  }

  String _ampmHour(int h) {
    if (h == 0) return '12 AM';
    if (h < 12) return '$h AM';
    if (h == 12) return '12 PM';
    return '${h - 12} PM';
  }

  String _formatTimes(String start, int durationMin) {
    if (start.isEmpty) return '';
    final parts = start.split(':');
    final h = int.tryParse(parts.first);
    final m = parts.length > 1 ? int.tryParse(parts[1]) : 0;
    if (h == null) return '';
    String hh(int hour, int minute) {
      final a = hour % 12 == 0 ? 12 : hour % 12;
      final mm = minute.toString().padLeft(2, '0');
      return '$a:$mm ${hour < 12 ? 'AM' : 'PM'}';
    }

    if (durationMin <= 0) return hh(h, m ?? 0);
    final end = h * 60 + (m ?? 0) + durationMin;
    return '${hh(h, m ?? 0)} – ${hh(end ~/ 60, end % 60)}';
  }

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final chosenDays = days ?? (sessionType.toLowerCase().contains('weekend') ? [6, 7] : [1, 2, 3, 4, 5]);
    final scheduled =
        entries.where((e) => chosenDays.contains(e.dayOfWeek) && e.start.isNotEmpty).toList();
    final offDays = entries.where((e) => !chosenDays.contains(e.dayOfWeek) && e.start.isNotEmpty).toList();

    var lo = 8, hi = 19;
    for (final e in scheduled) {
      final h = _hourOf(e.start);
      lo = min(lo, h);
      hi = max(hi, h + _span(e.durationMin));
    }
    lo = lo.clamp(0, 23);
    hi = hi.clamp(lo + 1, 24);
    final hours = List<int>.generate(hi - lo, (i) => i + lo);

    final today = DateTime.now().weekday; // DateTime: Mon=1…Sun=7 — same as the days table.

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        SingleChildScrollView(
          scrollDirection: Axis.horizontal,
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              // Time column.
              SizedBox(
                width: _timeW,
                child: Column(
                  children: [
                    Container(
                      height: 26,
                      alignment: Alignment.topLeft,
                      padding: const EdgeInsets.only(left: 4, top: 4),
                      child: Text(
                        'Time',
                        style: TextStyle(fontSize: 10, fontWeight: FontWeight.bold, color: scheme.primary),
                      ),
                    ),
                    for (final h in hours)
                      SizedBox(
                        height: _rowH,
                        child: Padding(
                          padding: const EdgeInsets.only(top: 2),
                          child: Text(
                            _ampmHour(h),
                            style: TextStyle(fontSize: 9, color: scheme.onSurfaceVariant),
                          ),
                        ),
                      ),
                  ],
                ),
              ),
              for (final d in chosenDays) _dayColumn(context, d, hours, scheduled, today == d),
            ],
          ),
        ),
        if (offDays.isNotEmpty) ...[
          const SizedBox(height: 12),
          Text(
            "Not on this week's days",
            style: TextStyle(fontSize: 12, fontWeight: FontWeight.bold, color: scheme.tertiary),
          ),
          for (final e in offDays)
            Text(
              '• ${e.name} — ${kDays[e.dayOfWeek]} ${e.start}'.trim(),
              style: TextStyle(fontSize: 12, color: scheme.onSurfaceVariant),
            ),
        ],
        if (unscheduledNames.isNotEmpty) ...[
          const SizedBox(height: 12),
          Text(
            'Not yet scheduled',
            style: TextStyle(fontSize: 12, fontWeight: FontWeight.bold, color: scheme.tertiary),
          ),
          for (final n in unscheduledNames)
            Text('• $n', style: TextStyle(fontSize: 12, color: scheme.onSurfaceVariant)),
        ],
      ],
    );
  }

  Widget _dayColumn(
    BuildContext context,
    int day,
    List<int> hours,
    List<TtEntry> scheduled,
    bool isToday,
  ) {
    final scheme = Theme.of(context).colorScheme;
    return Column(
      children: [
        Container(
          width: _dayW,
          height: 26,
          alignment: Alignment.centerLeft,
          padding: const EdgeInsets.only(left: 4),
          decoration: BoxDecoration(color: isToday ? scheme.primary : scheme.surfaceContainerHighest),
          child: Text(
            isToday ? '${kDays[day]} •' : kDays[day],
            style: TextStyle(
              fontSize: 10,
              fontWeight: FontWeight.bold,
              color: isToday ? scheme.onPrimary : scheme.onSurfaceVariant,
            ),
          ),
        ),
        SizedBox(
          width: _dayW,
          height: _rowH * hours.length,
          child: Stack(
            children: [
              for (var i = 0; i < hours.length; i++)
                Positioned(
                  left: 0,
                  right: 0,
                  top: _rowH * i,
                  height: _rowH,
                  child: Container(
                    decoration: BoxDecoration(
                      border: Border(bottom: BorderSide(color: scheme.outlineVariant)),
                    ),
                  ),
                ),
              for (final e in scheduled.where((s) => s.dayOfWeek == day))
                Positioned(
                  left: 0,
                  right: 0,
                  top: _rowH * (_hourOf(e.start) - hours.first),
                  height: _rowH * _span(e.durationMin),
                  child: Padding(
                    padding: const EdgeInsets.all(1.5),
                    child: Container(
                      decoration: BoxDecoration(
                        color: scheme.surface,
                        borderRadius: BorderRadius.circular(5),
                        border: Border.all(color: scheme.primary),
                      ),
                      padding: const EdgeInsets.all(2),
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          Text(
                            e.name,
                            maxLines: 2,
                            overflow: TextOverflow.ellipsis,
                            style: TextStyle(
                              fontSize: 10,
                              height: 1.1,
                              fontWeight: FontWeight.w600,
                              color: scheme.primary,
                            ),
                          ),
                          if (e.durationMin > 0)
                            Text(
                              _formatTimes(e.start, e.durationMin),
                              maxLines: 1,
                              overflow: TextOverflow.ellipsis,
                              style: TextStyle(fontSize: 8, height: 1.2, color: scheme.onSurfaceVariant),
                            ),
                          if (e.detail.isNotEmpty)
                            Text(
                              e.detail,
                              maxLines: _span(e.durationMin) > 1 ? 3 : 1,
                              overflow: TextOverflow.ellipsis,
                              style: TextStyle(fontSize: 8, height: 1.2, color: scheme.onSurfaceVariant),
                            ),
                        ],
                      ),
                    ),
                  ),
                ),
            ],
          ),
        ),
      ],
    );
  }
}